librz/bin/p/bin_elf.inc

```diff
@@ -299,391 +299,204 @@ static RzList *maps(RzBinFile *bf) {
-
-	if (!obj || !(ret = rz_list_newf((RzListFree)rz_bin_section_free))) {
-		return NULL;
+static ut64 reloc_target_size(struct Elf_(rz_bin_elf_obj_t) * obj) {
+	ut64 r = Elf_(rz_bin_elf_get_bits)(obj);
+	if (!r) {
+		return 8;
 	}
+	return r / 8;
+}
 
-	//there is not leak in section since they are cached by elf.c
-	//and freed within Elf_(rz_bin_elf_free)
-	if ((section = Elf_(rz_bin_elf_get_sections)(obj))) {
-		for (i = 0; !section[i].last; i++) {
-			if (!(ptr = RZ_NEW0(RzBinSection))) {
-				break;
-			}
-			ptr->name = strdup((char *)section[i].name);
-			if (strstr(ptr->name, "data") && !strstr(ptr->name, "rel") && !strstr(ptr->name, "pydata")) {
-				ptr->is_data = true;
-			} else if (!strcmp(ptr->name, "C")) {
-				ptr->is_data = true;
-			}
-			if (is_wordable_section(ptr->name)) {
-				ptr->format = rz_str_newf("Cd %d[%" PFMT64d "]",
-					RZ_BIN_ELF_WORDSIZE, section[i].size / RZ_BIN_ELF_WORDSIZE);
+/// base vaddr where to map the artificial reloc target vfile
+static ut64 reloc_targets_map_base(RzBinFile *bf, struct Elf_(rz_bin_elf_obj_t) * obj) {
+	if (obj->reloc_targets_map_base_calculated) {
+		return obj->reloc_targets_map_base;
+	}
+	// find the lowest unmapped address
+	ut64 max = 0;
+	RzList *maps = maps_unpatched(bf);
+	if (maps) {
+		RzListIter *it;
+		RzBinMap *map;
+		rz_list_foreach (maps, it, map) {
+			ut64 addr = map->vaddr + map->vsize;
+			if (addr > max) {
+				max = addr;
 			}
-			ptr->size = section[i].type != SHT_NOBITS ? section[i].size : 0;
-			ptr->vsize = section[i].size;
-			ptr->paddr = section[i].offset;
-			ptr->vaddr = section[i].rva;
-			ptr->type = section[i].type;
-			ptr->flags = section[i].flags;
-			ptr->perm = section_perms_from_flags(section[i].flags);
-			rz_list_append(ret, ptr);
 		}
+		rz_list_free(maps);
 	}
+	obj->reloc_targets_map_base_calculated = true;
+	max += 0x8; // small additional shift to not overlap with symbols like _end
+	obj->reloc_targets_map_base = max + rz_num_align_delta(max, reloc_target_size(obj));
+	return obj->reloc_targets_map_base;
+}
 
-	// program headers is another section
-	ut16 mach = obj->ehdr.e_machine;
-	num = obj->ehdr.e_phnum;
-	phdr = obj->phdr;
-	if (phdr) {
-		int n = 0;
-		for (i = 0; i < num; i++) {
-			if (!(ptr = RZ_NEW0(RzBinSection))) {
-				return ret;
-			}
-			ptr->size = phdr[i].p_filesz;
-			ptr->vsize = phdr[i].p_memsz;
-			ptr->paddr = phdr[i].p_offset;
-			ptr->vaddr = phdr[i].p_vaddr;
-			ptr->perm = phdr[i].p_flags;
-			ptr->align = phdr[i].p_align;
-			ptr->is_segment = true;
-			switch (phdr[i].p_type) {
-			case PT_DYNAMIC:
-				ptr->name = strdup("DYNAMIC");
+/// size of the artificial reloc target vfile
+static ut64 reloc_targets_vfile_size(RzBinFile *bf, struct Elf_(rz_bin_elf_obj_t) * obj) {
+	if (!bf->o || !bf->o->opts.patch_relocs) {
+		return 0;
+	}
+	return obj->g_reloc_num * reloc_target_size(obj);
+}
+
+static void patch_reloc(struct Elf_(rz_bin_elf_obj_t) * obj, RzBinElfReloc *rel, ut64 S, ut64 B, ut64 L) {
```

Comment:
I wonder if it makes sense to move it into `librz/bin/format/elf` directory and call `elf_reloc.c` since it might grow quite big if we have to support all reloc types.

Reply:
That is a good idea, the `patch_relocs` and `patch_reloc` could indeed fit well in `librz/bin/format/elf`. Though I just tried doing this refactor and it turned out not entirely trivial because `reloc_targets_map_base()` uses `maps_unpatched()` so I think we should postpone it for after this pr.

Reply:
Opened an issue: https://github.com/rizinorg/rizin/issues/1141